File: _engine/go/v0.1.0/mock.go
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

package engine

func FindMockOutput(entries []MockEntry, inputData map[string]any) map[string]any {
	var fallback map[string]any

	for _, entry := range entries {
		if entry.Input == nil {
			if fallback == nil {
				fallback = entry.Output
			}
			continue
		}
		if deepEqual(entry.Input, inputData) {
			return entry.Output
		}
	}

	return fallback
}

func GenerateFromSchema(schema map[string]any) any {
	if def, ok := schema["default"]; ok {
		return def
	}
	if enumVal, ok := schema["enum"]; ok {
		if arr, ok := enumVal.([]any); ok && len(arr) > 0 {
			return arr[0]
		}
	}

	schemaType := ""
	if t, ok := schema["type"]; ok {
		switch v := t.(type) {
		case string:
			schemaType = v
		case []any:
			if len(v) > 0 {
				if s, ok := v[0].(string); ok {
					schemaType = s
				}
			}
		}
	}

	switch schemaType {
	case "string":
		if desc, ok := schema["description"].(string); ok {
			return desc
		}
		return ""
	case "number", "integer":
		return float64(0)
	case "boolean":
		return false
	case "null":
		return nil
	case "array":
		if items, ok := schema["items"].(map[string]any); ok {
			return []any{GenerateFromSchema(items)}
		}
		return []any{}
	case "object":
		obj := map[string]any{}
		if props, ok := schema["properties"].(map[string]any); ok {
			for key, propSchema := range props {
				if ps, ok := propSchema.(map[string]any); ok {
					obj[key] = GenerateFromSchema(ps)
				}
			}
		}
		return obj
	default:
		return nil
	}
}

File: _engine/python/v0_2_0/mock.py
# Copyright 2026 GildenEye
# SPDX-License-Identifier: Apache-2.0

from typing import Any
from .types import MockEntry


def find_mock_output(
    mock_data: list[MockEntry], input_data: dict[str, Any]
) -> dict[str, Any] | None:
    fallback: dict[str, Any] | None = None

    for entry in mock_data:
        if entry.input is None and fallback is None:
            fallback = entry.output
            continue

        if entry.input is not None and entry.input == input_data:
            return entry.output

    return fallback


def generate_from_schema(schema: dict[str, Any]) -> Any:
    if "default" in schema:
        return schema["default"]
    if "enum" in schema and schema["enum"]:
        return schema["enum"][0]

    schema_type = schema.get("type")
    if isinstance(schema_type, list):
        schema_type = schema_type[0]

    if schema_type == "string":
        return schema.get("description", "")
    elif schema_type in ("number", "integer"):
        return 0
    elif schema_type == "boolean":
        return False
    elif schema_type == "null":
        return None
    elif schema_type == "array":
        items = schema.get("items")
        return [generate_from_schema(items)] if items else []
    elif schema_type == "object":
        obj: dict[str, Any] = {}
        properties = schema.get("properties", {})
        for key, prop_schema in properties.items():
            obj[key] = generate_from_schema(prop_schema)
        return obj
    else:
        return None
